File: Daily Questions/1568-minimum-number-of-days-to-disconnect-island/1568-minimum-number-of-days-to-disconnect-island.c
void dfs(int** grid, int gridSize, int gridColSize, int i, int j, bool** visited) {
    if (i < 0 || i >= gridSize || j < 0 || j >= gridColSize || grid[i][j] == 0 || visited[i][j])
        return;

    visited[i][j] = true;

    dfs(grid, gridSize, gridColSize, i + 1, j, visited);
    dfs(grid, gridSize, gridColSize, i - 1, j, visited);
    dfs(grid, gridSize, gridColSize, i, j + 1, visited);
    dfs(grid, gridSize, gridColSize, i, j - 1, visited);
}

bool isConnected(int** grid, int gridSize, int gridColSize) {
    bool** visited = (bool**)malloc(gridSize * sizeof(bool*));
    for (int i = 0; i < gridSize; i++) {
        visited[i] = (bool*)calloc(gridColSize, sizeof(bool));
    }

    bool found_island = false;
    for (int i = 0; i < gridSize; i++) {
        for (int j = 0; j < gridColSize; j++) {
            if (grid[i][j] == 1 && !visited[i][j]) {
                if (found_island) {
                    // More than one island found, no need to check further
                    for (int k = 0; k < gridSize; k++) {
                        free(visited[k]);
                    }
                    free(visited);
                    return false;
                }
                dfs(grid, gridSize, gridColSize, i, j, visited);
                found_island = true;
            }
        }
    }

    for (int i = 0; i < gridSize; i++) {
        free(visited[i]);
    }
    free(visited);

    return found_island;
}

int countLandCells(int** grid, int gridSize, int gridColSize) {
    int count = 0;
    for (int i = 0; i < gridSize; i++) {
        for (int j = 0; j < gridColSize; j++) {
            if (grid[i][j] == 1) {
                count++;
            }
        }
    }
    return count;
}

int minDays(int** grid, int gridSize, int* gridColSize) {
    if (!isConnected(grid, gridSize, *gridColSize)) {
        return 0;
    }

    if (countLandCells(grid, gridSize, *gridColSize) == 1) {
        return 1;
    }

    for (int i = 0; i < gridSize; i++) {
        for (int j = 0; j < gridColSize[i]; j++) {
            if (grid[i][j] == 1) {
                grid[i][j] = 0;
                if (!isConnected(grid, gridSize, *gridColSize)) {
                    grid[i][j] = 1; // Restore the original state
                    return 1;
                }
                grid[i][j] = 1; // Restore the original state
            }
        }
    }

    return 2;
}

File: 1568-minimum-number-of-days-to-disconnect-island/1568-minimum-number-of-days-to-disconnect-island.py
class Solution:
    def minDays(self, grid: List[List[int]]) -> int:
        def is_connected():
            # Helper function to check if the grid is connected
            visited = [[False] * len(grid[0]) for _ in range(len(grid))]
            
            def dfs(i, j):
                if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == 0 or visited[i][j]:
                    return
                visited[i][j] = True
                dfs(i + 1, j)
                dfs(i - 1, j)
                dfs(i, j + 1)
                dfs(i, j - 1)
            
            found_island = False
            for i in range(len(grid)):
                for j in range(len(grid[0])):
                    if grid[i][j] == 1 and not visited[i][j]:
                        if found_island:
                            return False  # More than one island found
                        dfs(i, j)
                        found_island = True
            
            # If no island was found, it's disconnected
            return found_island

        def count_land_cells():
            return sum(grid[i][j] for i in range(len(grid)) for j in range(len(grid[0])))

        # If the grid is already disconnected
        if not is_connected():
            return 0
        
        # If the grid has only one land cell
        if count_land_cells() == 1:
            return 1
        
        # Try removing each land cell and check if it disconnects the grid
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    grid[i][j] = 0
                    if not is_connected():
                        return 1
                    grid[i][j] = 1
        
        # If no single removal causes disconnection, return 2
        return 2